src/transpile/fusion.hpp

```diff
@@ -260,7 +277,36 @@ double Fusion::get_cost(const op_t& op) const {
     return cost_factor_;
 }
 
-bool Fusion::aggregate_operations(oplist_t& ops, const int fusion_start, const int fusion_end) const {
+
+op_t Fusion::generate_fusion_operation(const std::vector<op_t>& fusioned_ops,
+                                       const reg_t &qubits,
+                                       bool use_superop_method) const {
+  // Run simulation
+  RngEngine dummy_rng;
+  ExperimentData dummy_data;
+
+  if (use_superop_method) {
+    QubitSuperoperator::State<> superop_simulator;
+    superop_simulator.initialize_qreg(qubits.size());
+    superop_simulator.apply_ops(fusioned_ops, dummy_data, dummy_rng);
+    return Operations::make_superop(qubits,
+                                    superop_simulator.qreg().matrix());
+  } else {
+    QubitUnitary::State<> unitary_simulator;
+    unitary_simulator.initialize_qreg(qubits.size());
+    unitary_simulator.apply_ops(fusioned_ops, dummy_data, dummy_rng);
+    return Operations::make_unitary(qubits,
+                                    unitary_simulator.qreg().matrix(),
+                                    std::string("fusion"));
+  }
+}
```

Comment:
I think this code can be made more generic to avoid code repetition. I've rewritten it using templates (see below). In this case it is a bit verbose because I've had to deal also with the `Operations::make...` which are themselves very specific.  Other problem is that, as we have most of the code in .hpp files, the free functions are exposed to everyone that includes the file. Although they can be put as private member functions in Fusion, I'd be careful not to start adding too many functions. I would normally put them in an anonymous namespace in the .cpp file.
Anyway, consider this a suggestion, as in this case both branches of the `if/else` are quite simple so feel free to keep it as it is now.

```
template <typename OPERATOR>
op_t make_fusion_op(const reg_t &qubits, OPERATOR& op_simulator);


template<> op_t make_fusion_op<QubitSuperoperator::State<>>(const reg_t &qubits,
                                                            QubitSuperoperator::State<>& op_simulator){
  return Operations::make_superop(qubits, op_simulator.qreg().matrix());
}


template<> op_t make_fusion_op<QubitUnitary::State<>>(const reg_t &qubits,
                                                      QubitUnitary::State<>& op_simulator){
  return Operations::make_unitary(qubits, op_simulator.qreg().matrix(),
                                  std::string("fusion"));
}

template <typename OPERATOR>
op_t get_fusion_operator(const std::vector<op_t>& fusioned_ops,
                         const reg_t &qubits) {
    RngEngine dummy_rng;
    ExperimentData dummy_data;
    OPERATOR op_simulator;
    op_simulator.initialize_qreg(qubits.size());
    op_simulator.apply_ops(fusioned_ops, dummy_data, dummy_rng);
    return make_fusion_op<OPERATOR>(qubits, op_simulator);
}

op_t Fusion::generate_fusion_operation(const std::vector<op_t>& fusioned_ops,
                                       const reg_t &qubits,
                                       bool use_superop_method) const {
  // Run simulation
  if (use_superop_method) {
      return get_fusion_operator<QubitSuperoperator::State<>>(fusioned_ops, qubits);
  }
  return get_fusion_operator<QubitUnitary::State<>>(fusioned_ops, qubits);
}
```

Reply:
Nice, this is  the way to do it.
One single comment, we used before in the codebase the `type_t` nomenclature for our 
types, so I'd recommend  using: `operator_t`

Reply:
I did it that way originally but ended up changing it because found the template specialization made the code harder to follow. I think with one slightly verbose function it is easier to understand whats going on. I also had some issues with template specialization giving compiler errors if you didn't have the specialization before the function that called it etc.

In future I would also like to be able to add a third instruction output that uses superoperator simulator but converts its output to Kraus that I don't think will fit nicely with template parameter specialization based on the simulator class.